npm3.5.2 removes deduped modules that are required by react-map-gl on systems that fail to install gl (optionalDependency of mapbox-gl-js 0.11.6)

If `mapbox-gl-js`'s optionalDependency `gl` fails to install, npm3 will remove its dependencies, even if they are already deduped and placed in the root node_modules tree. 

These erroneously removed dependencies are frequentry required by react-map-gl, and other dependencies of the root module. 

To reproduce, just install `react-map-gl` with `npm3.5.2` on a system where gl will fail to install:

```sh
$ npm i --save react-map-gl
WARN engine react-map-gl@0.6.1: wanted: {"node":"0.10.x","npm":"2.x"} (current: {"node":"0.10.32","npm":"3.5.2"})

...

> gl@2.1.5 install /home/udocker/allen-key/node_modules/gl
> prebuild --download

prebuild WARN install libX11.so.6: cannot open shared object file: No such file or directory
gyp: angle/build/standalone.gypi not found (cwd: /home/udocker/allen-key/node_modules/gl) while reading includes of angle/src/angle.gyp while loading dependencies of binding.gyp while trying to load binding.gyp
prebuild ERR! configure error
prebuild ERR! stack Error: `gyp` failed with exit code: 1
prebuild ERR! stack     at ChildProcess.onCpExit (/home/udocker/allen-key/node_modules/node-gyp/lib/configure.js:305:16)
prebuild ERR! stack     at ChildProcess.emit (events.js:98:17)
prebuild ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:810:12)
prebuild ERR! not ok
prebuild ERR! build Error: `gyp` failed with exit code: 1
prebuild ERR! build     at ChildProcess.onCpExit (/home/udocker/allen-key/node_modules/node-gyp/lib/configure.js:305:16)
prebuild ERR! build     at ChildProcess.emit (events.js:98:17)
prebuild ERR! build     at Process.ChildProcess._handle.onexit (child_process.js:810:12)
npm WARN install:gl@2.1.5 gl@2.1.5 install: `prebuild --download`
npm WARN install:gl@2.1.5 Exit status 2
```

Then verify your tree:

```bash
$ npm ls

npm ERR! missing: core-util-is@~1.0.0, required by readable-stream@1.1.13
npm ERR! missing: string_decoder@~0.10.x, required by readable-stream@1.1.13
npm ERR! missing: core-util-is@~1.0.0, required by readable-stream@1.0.33
npm ERR! missing: string_decoder@~0.10.x, required by readable-stream@1.0.33
npm ERR! missing: core-util-is@~1.0.0, required by readable-stream@2.0.5
npm ERR! missing: string_decoder@~0.10.x, required by readable-stream@2.0.5
npm ERR! missing: once@^1.3.0, required by glob@4.0.6
npm ERR! missing: core-util-is@~1.0.0, required by readable-stream@1.1.13
npm ERR! missing: string_decoder@~0.10.x, required by readable-stream@1.1.13
```

--- a/package.json
+++ b/package.json
@@ -22,7 +22,7 @@
     "d3": "^3.5.5",
     "global": "^4.3.0",
     "keymirror": "^0.1.1",
-    "mapbox-gl": "0.11.4",
+    "mapbox-gl": "0.12.2",
     "nop": "^1.0.0",
     "object-assign": "^3.0.0",
     "svg-transform": "0.0.1",

